hail/python/hailtop/aiogoogle/client/storage_client.py

```diff
@@ -652,7 +652,14 @@ async def _listfiles_recursive(self, bucket: str, name: str) -> AsyncIterator[Fi
             items = page.get('items')
             if items is not None:
                 for item in page['items']:
-                    yield GoogleStorageFileListEntry(f'gs://{bucket}/{item["name"]}', item)
+                    url = f'gs://{bucket}/{item["name"]}'
+                    if url.endswith('/') and item is not None:
```

Comment:
`item` cannot be None: we subscript it in the line above.

I'd write this like this:

```
  entry = GoogleStorageFileListEntry(url, item)
  if url.endswith('/'):
    status = await entry.status()
    size = await entry.size()
    if size != 0:
      raise FileAndDirectoryError(url)
  yield entry
```

The entry caches the status, so this way, if the user gets the status, it reuses the status object you create.